marlowe-runtime/test/Language/Marlowe/Runtime/Transaction/BuildConstraintsSpec.hs

```diff
@@ -223,3 +236,86 @@ withdrawSpec = Hspec.describe "buildWithdrawConstraints" do
+        , eraParams
+        }
+      unboundedEraSummary i = EraSummary
+        { eraStart = oneSecondBound i
+        , eraEnd = EraUnbounded
+        , eraParams
+        }
+      eraHistory =
+        EraHistory CardanoMode
+          . mkInterpreter
+          . summaryWithExactly
+          $ Exactly
+          $  K (oneSecondEraSummary 0) -- Byron lasted 1 second
+          :* K (oneSecondEraSummary 1) -- Shelley lasted 1 second
+          :* K (oneSecondEraSummary 2) -- Allegra lasted 1 second
+          :* K (oneSecondEraSummary 3) -- Mary lasted 1 second
+          :* K (oneSecondEraSummary 4) -- Alonzo lasted 1 second
+          :* K (unboundedEraSummary 5) -- Babbage never ends
+          :* Nil
+    Hspec.QuickCheck.prop "valid slot interval for timed-out contract" \assets utxo -> do
+      address <- arbitrary
+      -- The choice intervals for the tip, minimum time, and timeout overlap, so every ordering will occur.
+      tipSlot' <- chooseInteger (9, 20) -- Make sure the tip is in the Babbage era.
+      minTime <- oneof
+                   [
+                     (1000 *) <$> chooseInteger (6, 20)  -- Even seconds, so there is a chance of collision.
+                   , chooseInteger (6000, 20000)         -- Milliseconds, so there is rounding off.
+                   ]
+      timeout <- oneof
+                   [
+                     (1000 *) <$> chooseInteger (6, 20)  -- Even seconds, so there is a chance of collision.
+                   , chooseInteger (6000, 20000)         -- Milliseconds, so there is rounding off.
+                   ]
+      marloweParams <- arbitrary
+      let
+        toSlot = (`div` 1000)
+        tipSlot = toEnum . fromEnum $ tipSlot'
+        tipTime = 1000 * tipSlot'
+        marloweState = Semantics.State AM.empty AM.empty AM.empty $ POSIXTime minTime
+        marloweContract = Semantics.When [] (POSIXTime timeout) Semantics.Close  -- This contract can only time out.
+        datum = Semantics.MarloweData{..}
+        marloweOutput = TransactionScriptOutput{..}
+        result =
+          buildApplyInputsConstraints
+            systemStart eraHistory MarloweV1
+            marloweOutput
+            tipSlot
+            (Chain.TransactionMetadata mempty) Nothing Nothing mempty
+      pure
+        case result of
+            -- A valid transaction will occur if the tip is not before the timeout.
+            Right _ -> toSlot timeout <= tipSlot'
+            -- A useless transaction will occur if the tip is before the timeout.
+            Left (ApplyInputsConstraintsBuildupFailed (MarloweComputeTransactionFailed "TEUselessTransaction")) ->
+              tipTime < timeout
+            -- If rounding off causes the timeout to fall at the tip, then there is an ambiguity.
+            Left (ApplyInputsConstraintsBuildupFailed (MarloweComputeTransactionFailed "TEAmbiguousTimeIntervalError")) ->
+              tipSlot' == toSlot timeout && minTime <= timeout
```

Comment:
Wouldn't this also be the case if `tipSlot' >= timeout`? I guess it would fail if I was right, but I don't get it.

Reply:
This is outdated. I rewrote the test logic so it is cleaner and more transparent.